File: leetcode/go/algo/002.go
package algo

func AddTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	num := 0
	from_low := 0
	var tmp, head *ListNode

	for l1 != nil || l2 != nil {
		num=from_low
		if l1 != nil {
			num += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			num += l2.Val;
			l2 = l2.Next
		}
		if num > 9 {
			num = num % 10
			from_low = 1
		} else {
			from_low = 0
		}

		if nil == head {
			tmp = &ListNode{Val: num}
			head = tmp

		} else {
			tmp.Next = &ListNode{Val: num}
			tmp = tmp.Next
		}

	}
	if from_low > 0 {
		tmp.Next = &ListNode{Val: from_low}
	}
	return head
}

File: leetcode/php/002.php
<?php

/**
 * Created by PhpStorm.
 * User: 764432054@qq.com
 * Date: 2020/9/8
 * Time: 19:36
 */
include "include/common.php";

function addTwoNumbers($l1, $l2)
{
    $fromLow = 0;
    while ($l1 || $l2) {
        $num = $fromLow;
        if ($l1) {
            $num += $l1->val;
            $l1 = $l1->next;
        }
        if ($l2) {
            $num += $l2->val;
            $l2 = $l2->next;
        }
        if ($num > 9) {
            $fromLow = 1;
            $num = $num % 10;
        } else {
            $fromLow = 0;
        }
        if (!isset($head)) {
            $head = new ListNode($num);
            $tmp = $head;
        } else {
            $tmp->next = new ListNode($num);
            $tmp = $tmp->next;
        }
    }
    //比l1与l2都长 且有进位
    if ($fromLow > 0) {
        $tmp->next = new ListNode($fromLow);
    }
    return $head;
}
function addTwoNumbers1($l1, $l2) {
    $fromLow=0;
    $node=new ListNode(null);
    $cur=$node;
    while($l1 ||$l2){
        $fromLow=$fromLow>9?1:0;
        if ($l1) {
            $fromLow += $l1->val;
            $l1 = $l1->next;
        }
        if ($l2) {
            $fromLow += $l2->val;
            $l2 = $l2->next;
        }
        $now=$fromLow%10;
        $curNode=new ListNode($now);
        $cur->next=$curNode;
        $cur=$cur->next;
    }
    if($fromLow>9){
        $curNode=new ListNode(1);
        $cur->next=$curNode;
    }
    return $node->next;
}

$l1 = createNodeList([1, 3, 6]);
$l2 = createNodeList([2, 7, 5]);

print_r(addTwoNumbers($l1, $l2));


